### ISSUE TITLE
createdAtStart Koin flag is ignored

### ISSUE THREAD

**Issue by derkalaender** [Kotlin-Discord/kord-extensions#78]
# Description

Creating a Koin module or definition inside of the `beforeKoinSetup` hook, the `createdAtStart` flag gets ignored, as those modules/definitions do not in fact get instantiated at start. This flag is useful for Database connections for example, which need to be established directly at application start to do migrations etc.

This might not be a deal-breaker at first, since you could just instantiate the object yourself (if it doesnt inject anything) and then create a `single` from that. However, things get difficult when that object itself has other dependencies. So you'd either need to use getKoin() manually or create those other dependencies outside of Koin as well, which might get ugly pretty quickly.

# Versions

Latest Kord Extensions `1.5.0-SNAPSHOT`

# Code Examples

```kotlin
class MyDatabase(dep: DepOne) : KoinComponent, KLoggable {
    override val logger: KLogger = logger()
    
    val otherDep: DepTwo by inject()

    init {
        logger.info { "Setting up database with dependencies $dep and $otherDep" }
    }
}

suspend fun ExtensibleBotBuilder.setupDatabase() {
    hooks {
        beforeKoinSetup {
            loadModule {
                single(createdAtStart = true) { MyDatabase(get()) }
            }
        }
    }
}
```
No log can be seen (only once another object uses MyDatabase). MyDatabase can also not be instantiated externally because of the `inject`

# Suggestions

[Here](https://github.com/Kord-Extensions/kord-extensions/blob/ff5eab0549dd094e3cfbcd1b45400c951e6f32e5/kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/builders/ExtensibleBotBuilder.kt#L212-L224) in setupKoin() the startKoin() is called before running the runBeforeKoinSetup hooks. This might be the culprit, because according to [Koin docs](https://insert-koin.io/docs/reference/koin-core/definitions/#create-instances-at-start), anything flagged with createdAtStart
is executed once startKoin is called and at that point, the modules have not been registered yet.

**Comment by gdude2002** [comment]
Hello, just getting around to this now!

Do you know if modules can be registered globally before `startKoin` is called? I guess the implication here is that they can, but it'd be good to double-check whether `getKoin` is safe to use that early.

I'll look into it.

**Comment by gdude2002** [comment]
Nope, `loadKoinModules()` can't be called before `startKoin` - any ideas?

**Comment by gdude2002** [comment]
Marking this stale as it hasn't been updated in 5 days.

**Comment by gdude2002** [comment]
This hasn't been updated for 10 days, and I'm at a loss as to how this could be implemented. As the reporter seems to have disappeared, I'll be closing this issue now - if you have any further ideas or a proof of concept, please feel free to open another issue or comment on this one.
